### PR TITLE
[MessageBoard] Add TS Demo for Message board

### PR DESCRIPTION
**Pull request by areshand**
## Motivation

1. Demo the ACL messageboard with basic JS (**messageboard_demo/client/index.ts**)
2. Add Cli option to move_example for compiling move with named addresses
3. add ACL.move and Offer.move to messageboard since move nursery is not ready to be included in move stdlib yet. 

followup: (1) add demo for cap based board

### Have you read the [Contributing Guidelines on pull requests](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan
cargo test -p framework

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/aptos-labs/aptos-core/tree/main/developer-docs-site, and link to your PR here.)

### CODE REVIEW

File: developer-docs-site/static/move_demo/client/index.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
+const aptos = require("aptos");
```
Comment (CapCap) [inline developer-docs-site/static/move_demo/client/index.js:1]: could you use TS for this please, like some of the other demos?

Comment (areshand) [reply developer-docs-site/static/move_demo/client/index.js]: addressed the comments

File: aptos-move/move-examples/messageboard/sources/Offer.move [hunk 2]
```diff
@@ -0,0 +1,138 @@
+/// Provides a way to transfer structs from one account to another in two transactions.
```
Comment (davidiw) [inline aptos-move/move-examples/messageboard/sources/Offer.move:1]: I feel like this should just be part of the standard library if you want to incorporate it.

File: aptos-move/move-examples/messageboard/sources/ACL.move [hunk 3]
```diff
@@ -0,0 +1,46 @@
+/// Access control list (ACL) module. An ACL is a list of account addresses who
```
Comment (davidiw) [inline aptos-move/move-examples/messageboard/sources/ACL.move:1]: same, as offer, if we're going to recommend using this, we should just make it generally available. That would also reduce burden on the chain, no?

File: aptos-move/move-examples/messageboard/sources/Offer.move [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  spec module {
+    /// Returns true if the recipient is allowed to redeem `Offer<Offered>` at `offer_address`
+    /// and false otherwise.
+    fun is_allowed_recipient<Offered>(offer_addr: address, recipient: address): bool {
+      recipient == global<Offer<Offered>>(offer_addr).for || recipient == offer_addr
+    }
+  }
+
+
+
```
Comment (davidiw) [inline aptos-move/move-examples/messageboard/sources/Offer.move:137]: wow, so many trees lol

File: developer-docs-site/static/messageboard_demo/client/index.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        const payload = {
+            type: "script_function_payload",
+            function: `${funcName}`,
+            type_arguments: [],
+            arguments: args
+        };
+        return payload;
+    }
+
+    async createMessageboard(account: AptosAccount, modulePath: string) {
```
Comment (davidiw) [inline developer-docs-site/static/messageboard_demo/client/index.ts:88]: this should be two different steps....

Ideally we should have "installMessageboardModule" and "createMessageboard" separately.

installation happens only once, but there's no reason why we would ever need more than one instance of the module installed. Whereas Alice and Bob can use the same module but call "createMessageboard" using the same backend. Much like we did in the first move module tutorial.

**Review by davidiw — CHANGES_REQUESTED** [review]
Really great start to an interactive app. It might be great to specify a little bit more here and actually build out a slightly more full fledged app that:
* keeps track of events that have been seen, so they aren't pulled again
* supports remove participant
* keeps a list of historical participants -- maybe even all possible participants in the case of the capability and acl tables -- to support removal.

With that being said, we've done a lot of work on this already, so it may be fine to just clean it up a bit with a solid test and pause effort here for a little bit.

This demo does something dangerous, which it teaches the user to install a module each time they wish to instantiate a resource. The problem with installing the same module over and over again is that someone would inevitably have to build an app that can guess that the module is effectively the same just installed in different places.

Also the code moves around the hello blockchain code which is reference in the first move module tutorial but it doesn't update the tutorial.

File: developer-docs-site/static/messageboard_demo/client/index.ts [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    var participant = new AptosAccount();
+    await faucetClient.fundAccount(admin.address(), 5000);
+    await faucetClient.fundAccount(participant.address(), 5000);
+
+    var board = new Messageboard(BoardType.ACL, admin);
+
+    // compile the modules with the admin's address
+    chdir('../../../../aptos-move/move-examples');
+    exec(
+        `cargo run -- messageboard/sources/ -a MessageBoard -h ${admin.address().toString()}`,
```
Comment (davidiw) [inline developer-docs-site/static/messageboard_demo/client/index.ts:154]: this is really cool, we should update the documentation for first move module asap.

File: aptos-move/move-examples/messageboard/sources/ACL.move (hunk 3 above)
Comment (areshand) [reply aptos-move/move-examples/messageboard/sources/ACL.move]: added to stdlib

File: aptos-move/move-examples/messageboard/sources/Offer.move (hunk 2 above)
Comment (areshand) [reply aptos-move/move-examples/messageboard/sources/Offer.move]: added to stdlib

File: aptos-move/move-examples/messageboard/sources/Offer.move (hunk 4 above)
Comment (areshand) [reply aptos-move/move-examples/messageboard/sources/Offer.move]: removed

File: developer-docs-site/static/messageboard_demo/client/index.ts (hunk 5 above)
Comment (areshand) [reply developer-docs-site/static/messageboard_demo/client/index.ts]: split to two

**Comment by areshand** [comment]
> 



> Really great start to an interactive app. It might be great to specify a little bit more here and actually build out a slightly more full fledged app that:
> 
> * keeps track of events that have been seen, so they aren't pulled again
currently, the SDK doesn't support  passing page or limit parameter. 
will support this feature once SDK supports passing page or limit parameters. 
> * supports remove participant
added

> * keeps a list of historical participants -- maybe even all possible participants in the case of the capability and acl tables -- to support removal.
added some basic tracking of participants. will refine when adding cap based board

> 
> With that being said, we've done a lot of work on this already, so it may be fine to just clean it up a bit with a solid test and pause effort here for a little bit.
> 
> This demo does something dangerous, which it teaches the user to install a module each time they wish to instantiate a resource. The problem with installing the same module over and over again is that someone would inevitably have to build an app that can guess that the module is effectively the same just installed in different places.

Yea, great callout. separate it to two functions to not misleading the reader. 
> 
> Also the code moves around the hello blockchain code which is reference in the first move module tutorial but it doesn't update the tutorial.

hmm, seems the move file path is not used in code. seems the website is displaying correctly https://deploy-preview-546--aptos-developer-docs.netlify.app/tutorials/your-first-move-module. am i missing something here?

File: aptos-move/move-examples/src/main.rs [hunk 7]
```diff
@@ -1,28 +1,38 @@
...
+    } else {
+        BTreeMap::new()
+    };
 
     let build_config = move_package::BuildConfig {
         dev_mode: false,
         generate_abis: false,
         generate_docs: true,
         install_dir: args.output_path,
+        additional_named_addresses: named_address,
```
Comment (davidiw) [inline aptos-move/move-examples/src/main.rs:35]: @gregnazario fyi

**Review by davidiw — COMMENTED** [review]
this is really cool

1) update first move module tutorial due to new paths
2) use the cool new parameter to make it easier to use named addresses -- no more of that nonsensical update the Move.toml file
3) consider consulting with Greg to port the compilation to the Aptos CLI
4) Do we have any readme or guide for this?

code looks good to me -- but it would break the existing documentation for first move module.

File: aptos-move/move-examples/src/main.rs (hunk 7 above)
Comment (gregnazario) [reply aptos-move/move-examples/src/main.rs:35]: Yeah, this is already built into the CLI

Comment (davidiw) [reply aptos-move/move-examples/src/main.rs:35]: oh, I see that @areshand did: `MessageBoard = "_"`

**Comment by areshand** [comment]
@davidiw  updated the tutorial so it is not broken. for the readme, will add after this lands so that I can link to the code in the readme more easily

File: aptos-move/move-examples/hello_blockchain/Move.toml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+[package]
+name = "Examples"
+version = "0.0.0"
+
+[addresses]
+HelloBlockchain = "66301a58029b93bb46e93d344a79d5673a62c81f94a829b93d0e548a19269fdf"
```
Comment (davidiw) [inline aptos-move/move-examples/hello_blockchain/Move.toml:6]: can you set this to "_" like you did for message board? and maybe update the instructions so folks can compile this more easily

**Review by davidiw — APPROVED** [review]
woot woot

File: aptos-move/move-examples/hello_blockchain/Move.toml (hunk 8 above)
Comment (areshand) [reply aptos-move/move-examples/hello_blockchain/Move.toml]: modified

**Pull request closed without merging**